# Week 5 Assignment 

#### Predicting the gender of voice recordings using _K-nearest neighbor_

Marissa Berk

The data set _voice.csv_ was created to identify a voice as male or female, based upon acoustic properties of the voice and speech. The dataset consists of 3,168 recorded voice samples, collected from male and female speakers. The voice samples are pre-processed by acoustic analysis in R using the seewave and tuneR packages, with an analyzed frequency range of 0hz-280hz More information on this dataset can be found here: https://www.kaggle.com/primaryobjects/voicegender/home

The dataset is cleaned, split into a training and test set and the K-nearest neighbor algorithm is used to predict which recordings are from female speakers.

In [66]:
import seaborn as sns
import sklearn as sk
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split #We need this to split the data

In [67]:
df = pd.read_csv('voice.csv')
df.head(30) #let's take a look at the dataset


meanfreq        sd    median       Q25       Q75       IQR       skew  \
0   0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1   0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2   0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3   0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4   0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   
5   0.132786  0.079557  0.119090  0.067958  0.209592  0.141634   1.932562   
6   0.150762  0.074463  0.160106  0.092899  0.205718  0.112819   1.530643   
7   0.160514  0.076767  0.144337  0.110532  0.231962  0.121430   1.397156   
8   0.142239  0.078018  0.138587  0.088206  0.208587  0.120381   1.099746   
9   0.134329  0.080350  0.121451  0.075580  0.201957  0.126377   1.190368   
10  0.157021  0.071943  0.168160  0.101430  0.216740  0.115310   0.979442   
11  0.138551  0.077054  0.127527  0.087314  0.202739  0.115426   1.626770   
12  0.137343  0.080877  0.124263  0.083145  0.209227  0.126082   1.378728   
13  0.181225  0.060042  0.190953  0.128839  0.229532  0.100693   1.369430   
14  0.183115  0.066982  0.191233  0.129149  0.240152  0.111004   3.568104   
15  0.174272  0.069411  0.190874  0.115602  0.228279  0.112677   4.485038   
16  0.190846  0.065790  0.207951  0.132280  0.244357  0.112076   1.562304   
17  0.171247  0.074872  0.152807  0.122391  0.243617  0.121227   3.207170   
18  0.168346  0.074121  0.145618  0.115756  0.239824  0.124068   2.704335   
19  0.173631  0.073352  0.153569  0.123680  0.244234  0.120554   2.804975   
20  0.172754  0.076903  0.177736  0.120070  0.245368  0.125298   2.967765   
21  0.181015  0.074369  0.169299  0.128673  0.254175  0.125502   2.587325   
22  0.163536  0.072449  0.145543  0.113930  0.227449  0.113519   3.587650   
23  0.170213  0.075105  0.146053  0.123989  0.250126  0.126137   2.816793   
24  0.160422  0.076615  0.144824  0.120924  0.237244  0.116319   6.253208   
25  0.164700  0.075362  0.147018  0.118698  0.240475  0.121777   4.208608   
26  0.169579  0.075635  0.186468  0.116706  0.238549  0.121843   4.269923   
27  0.169021  0.071778  0.143168  0.125801  0.248315  0.122515   3.079273   
28  0.167340  0.072841  0.141739  0.122174  0.240000  0.117826   2.192126   
29  0.180528  0.070867  0.142385  0.129541  0.252477  0.122936   2.799969   

           kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0    274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1    634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2   1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3      4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4      4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   
5      8.308895  0.963181  0.738307  ...  0.132786  0.110132  0.017112   
6      5.987498  0.967573  0.762638  ...  0.150762  0.105945  0.026230   
7      4.766611  0.959255  0.719858  ...  0.160514  0.093052  0.017758   
8      4.070284  0.970723  0.770992  ...  0.142239  0.096729  0.017957   
9      4.787310  0.975246  0.804505  ...  0.134329  0.105881  0.019300   
10     3.974223  0.965249  0.733693  ...  0.157021  0.088894  0.022069   
11     6.291365  0.966004  0.752042  ...  0.138551  0.104199  0.019139   
12     5.008952  0.963514  0.736150  ...  0.137343  0.092644  0.016789   
13     5.475600  0.937446  0.537080  ...  0.181225  0.131504  0.025000   
14    35.384748  0.940333  0.571394  ...  0.183115  0.102799  0.020833   
15    61.764908  0.950972  0.635199  ...  0.174272  0.102046  0.018328   
16     7.834350  0.938546  0.538810  ...  0.190846  0.113323  0.017544   
17    25.765565  0.936954  0.586420  ...  0.171247  0.079718  0.015671   
18    18.484703  0.934523  0.559742  ...  0.168346  0.083484  0.015717   
19    20.857543  0.930917  0.518269  ...  0.173631  0.090130  0.015702   
20    20.078115  0.925539  0.523081  ...  0.172754  0.093574  

Now that we can see the dataset, let's take a look at how the variables correlate with eachother so we can select the variables we want to include in our prediction model.

In [68]:
df.corr()

meanfreq        sd    median       Q25       Q75       IQR  \
meanfreq  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
sd       -0.739039  1.000000 -0.562603 -0.846931 -0.161076  0.874660   
median    0.925445 -0.562603  1.000000  0.774922  0.731849 -0.477352   
Q25       0.911416 -0.846931  0.774922  1.000000  0.477140 -0.874189   
Q75       0.740997 -0.161076  0.731849  0.477140  1.000000  0.009636   
IQR      -0.627605  0.874660 -0.477352 -0.874189  0.009636  1.000000   
skew     -0.322327  0.314597 -0.257407 -0.319475 -0.206339  0.249497   
kurt     -0.316036  0.346241 -0.243382 -0.350182 -0.148881  0.316185   
sp.ent   -0.601203  0.716620 -0.502005 -0.648126 -0.174905  0.640813   
sfm      -0.784332  0.838086 -0.661690 -0.766875 -0.378198  0.663601   
mode      0.687715 -0.529150  0.677433  0.591277  0.486857 -0.403764   
centroid  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
meanfun   0.460844 -0.466281  0.414909  0.545035  0.155091 -0.534462   
minfun    0.383937 -0.345609  0.337602  0.320994  0.258002 -0.222680   
maxfun    0.274004 -0.129662  0.251328  0.199841  0.285584 -0.069588   
meandom   0.536666 -0.482726  0.455943  0.467403  0.359181 -0.333362   
mindom    0.229261 -0.357667  0.191169  0.302255 -0.023750 -0.357037   
maxdom    0.519528 -0.482278  0.438919  0.459683  0.335114 -0.337877   
dfrange   0.515570 -0.475999  0.435621  0.454394  0.335648 -0.331563   
modindx  -0.216979  0.122660 -0.213298 -0.141377 -0.216475  0.041252   

              skew      kurt    sp.ent       sfm      mode  centroid  \
meanfreq -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
sd        0.314597  0.346241  0.716620  0.838086 -0.529150 -0.739039   
median   -0.257407 -0.243382 -0.502005 -0.661690  0.677433  0.925445   
Q25      -0.319475 -0.350182 -0.648126 -0.766875  0.591277  0.911416   
Q75      -0.206339 -0.148881 -0.174905 -0.378198  0.486857  0.740997   
IQR       0.249497  0.316185  0.640813  0.663601 -0.403764 -0.627605   
skew      1.000000  0.977020 -0.195459  0.079694 -0.434859 -0.322327   
kurt      0.977020  1.000000 -0.127644  0.109884 -0.406722 -0.316036   
sp.ent   -0.195459 -0.127644  1.000000  0.866411 -0.325298 -0.601203   
sfm       0.079694  0.109884  0.866411  1.000000 -0.485913 -0.784332   
mode     -0.434859 -0.406722 -0.325298 -0.485913  1.000000  0.687715   
centroid -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
meanfun  -0.167668 -0.194560 -0.513194 -0.421066  0.324771  0.460844   
minfun   -0.216954 -0.203201 -0.305826 -0.362100  0.385467  0.383937   
maxfun   -0.080861 -0.045667 -0.120738 -0.192369  0.172329  0.274004   
meandom  -0.336848 -0.303234 -0.293562 -0.428442  0.491479  0.536666   
mindom   -0.061608 -0.103313 -0.294869 -0.289593  0.198150  0.229261   
maxdom   -0.305651 -0.274500 -0.324253 -0.436649  0.477187  0.519528   
dfrange  -0.304640 -0.272729 -0.319054 -0.431580  0.473775  0.515570   
modindx  -0.169325 -0.205539  0.198074  0.211477 -0.182344 -0.216979   

           meanfun    minfun    maxfun   meandom    mindom    maxdom  \
meanfreq  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
sd       -0.466281 -0.345609 -0.129662 -0.482726 -0.357667 -0.482278   
median    0.414909  0.337602  0.251328  0.455943  0.191169  0.438919   
Q25       0.545035  0.320994  0.199841  0.467403  0.302255  0.459683   
Q75       0.155091  0.258002  0.285584  0.359181 -0.023750  0.335114   
IQR      -0.534462 -0.222680 -0.069588 -0.333362 -0.357037 -0.337877   
skew     -0.167668 -0.216954 -0.080861 -0.336848 -0.061608 -0.305651   
kurt     -0.194560 -0.203201 -0.045667 -0.303234 -0.103313 -0.274500   
sp.ent   -0.513194 -0.305826 -0.120738 -0.293562 -0.294869 -0.324253   
sfm      -0.421066 -0.362100 -0.192369 -0.428442 -0.289593 -0.436649   
mode      0.324771  0.385467  0.172329  0.491479  0.198150  0.477187   
centroid  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
meanfun   1.000000  0.339387  0.311950  0.270840

Unfortunately the variable we need: _label_ does not show up in this correlation table so we need to make dummie variables for the _label_ variable.

# Data Cleaning & Variable Selection

In [69]:
dummies = pd.get_dummies(df['label'])#make dummie variables for the gender
dummies
df = pd.concat([df, dummies], axis=1) #the axis=1 means: add it to the columns
df.head(10)

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   
5  0.132786  0.079557  0.119090  0.067958  0.209592  0.141634   1.932562   
6  0.150762  0.074463  0.160106  0.092899  0.205718  0.112819   1.530643   
7  0.160514  0.076767  0.144337  0.110532  0.231962  0.121430   1.397156   
8  0.142239  0.078018  0.138587  0.088206  0.208587  0.120381   1.099746   
9  0.134329  0.080350  0.121451  0.075580  0.201957  0.126377   1.190368   

          kurt    sp.ent       sfm  ...    minfun    maxfun   meandom  \
0   274.402906  0.893369  0.491918  ...  0.015702  0.275862  0.007812   
1   634.613855  0.892193  0.513724  ...  0.015826  0.250000  0.009014   
2  1024.927705  0.846389  0.478905  ...  0.015656  0.271186  0.007990   
3     4.177296  0.963322  0.727232  ...  0.017798  0.250000  0.201497   
4     4.333713  0.971955  0.783568  ...  0.016931  0.266667  0.712812   
5     8.308895  0.963181  0.738307  ...  0.017112  0.253968  0.298222   
6     5.987498  0.967573  0.762638  ...  0.026230  0.266667  0.479620   
7     4.766611  0.959255  0.719858  ...  0.017758  0.144144  0.301339   
8     4.070284  0.970723  0.770992  ...  0.017957  0.250000  0.336476   
9     4.787310  0.975246  0.804505  ...  0.019300  0.262295  0.340365   

     mindom    maxdom   dfrange   modindx  label  female  male  
0  0.007812  0.007812  0.000000  0.000000   male       0     1  
1  0.007812  0.054688  0.046875  0.052632   male       0     1  
2  0.007812  0.015625  0.007812  0.046512   male       0     1  
3  0.007812  0.562500  0.554688  0.247119   male       0     1  
4  0.007812  5.484375  5.476562  0.208274   male       0     1  
5  0.007812  2.726562  2.718750  0.125160   male       0     1  
6  0.007812  5.312500  5.304688  0.123992   male       0     1  
7  0.007812  0.539062  0.531250  0.283937   male       0     1  
8  0.007812  2.164062  2.156250  0.148272   male       0     1  
9  0.015625  4.695312  4.679688  0.089920   male       0     1  

[10 rows x 23 columns]

Now there are two new variables- female and male. Let's take a look at the correlation matrix now.

In [45]:
df.corr()

meanfreq        sd    median       Q25       Q75       IQR  \
meanfreq  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
sd       -0.739039  1.000000 -0.562603 -0.846931 -0.161076  0.874660   
median    0.925445 -0.562603  1.000000  0.774922  0.731849 -0.477352   
Q25       0.911416 -0.846931  0.774922  1.000000  0.477140 -0.874189   
Q75       0.740997 -0.161076  0.731849  0.477140  1.000000  0.009636   
IQR      -0.627605  0.874660 -0.477352 -0.874189  0.009636  1.000000   
skew     -0.322327  0.314597 -0.257407 -0.319475 -0.206339  0.249497   
kurt     -0.316036  0.346241 -0.243382 -0.350182 -0.148881  0.316185   
sp.ent   -0.601203  0.716620 -0.502005 -0.648126 -0.174905  0.640813   
sfm      -0.784332  0.838086 -0.661690 -0.766875 -0.378198  0.663601   
mode      0.687715 -0.529150  0.677433  0.591277  0.486857 -0.403764   
centroid  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
meanfun   0.460844 -0.466281  0.414909  0.545035  0.155091 -0.534462   
minfun    0.383937 -0.345609  0.337602  0.320994  0.258002 -0.222680   
maxfun    0.274004 -0.129662  0.251328  0.199841  0.285584 -0.069588   
meandom   0.536666 -0.482726  0.455943  0.467403  0.359181 -0.333362   
mindom    0.229261 -0.357667  0.191169  0.302255 -0.023750 -0.357037   
maxdom    0.519528 -0.482278  0.438919  0.459683  0.335114 -0.337877   
dfrange   0.515570 -0.475999  0.435621  0.454394  0.335648 -0.331563   
modindx  -0.216979  0.122660 -0.213298 -0.141377 -0.216475  0.041252   
female    0.337415 -0.479539  0.283919  0.511455 -0.066906 -0.618916   
male     -0.337415  0.479539 -0.283919 -0.511455  0.066906  0.618916   

              skew      kurt    sp.ent       sfm  ...   meanfun    minfun  \
meanfreq -0.322327 -0.316036 -0.601203 -0.784332  ...  0.460844  0.383937   
sd        0.314597  0.346241  0.716620  0.838086  ... -0.466281 -0.345609   
median   -0.257407 -0.243382 -0.502005 -0.661690  ...  0.414909  0.337602   
Q25      -0.319475 -0.350182 -0.648126 -0.766875  ...  0.545035  0.320994   
Q75      -0.206339 -0.148881 -0.174905 -0.378198  ...  0.155091  0.258002   
IQR       0.249497  0.316185  0.640813  0.663601  ... -0.534462 -0.222680   
skew      1.000000  0.977020 -0.195459  0.079694  ... -0.167668 -0.216954   
kurt      0.977020  1.000000 -0.127644  0.109884  ... -0.194560 -0.203201   
sp.ent   -0.195459 -0.127644  1.000000  0.866411  ... -0.513194 -0.305826   
sfm       0.079694  0.109884  0.866411  1.000000  ... -0.421066 -0.362100   
mode     -0.434859 -0.406722 -0.325298 -0.485913  ...  0.324771  0.385467   
centroid -0.322327 -0.316036 -0.601203 -0.784332  ...  0.460844  0.383937   
meanfun  -0.167668 -0.194560 -0.513194 -0.421066  ...  1.000000  0.339387   
minfun   -0.216954 -0.203201 -0.305826 -0.362100  ...  0.339387  1.000000   
maxfun   -0.080861 -0.045667 -0.120738 -0.192369  ...  0.311950  0.213987   
meandom  -0.336848 -0.303234 -0.293562 -0.428442  ...  0.270840  0.375979   
mindom   -0.061608 -0.103313 -0.294869 -0.289593  ...  0.162163  0.082015   
maxdom   -0.305651 -0.274500 -0.324253 -0.436649  ...  0.277982  0.317860   
dfrange  -0.304640 -0.272729 -0.319054 -0.431580  ...  0.275154  0.316486   
modindx  -0.169325 -0.205539  0.198074  0.211477  ... -0.054858  0.002042   
female   -0.036627 -0.087195 -0.490552 -0.357499  ...  0.833921  0.136692   
male      0.036627  0.087195  0.490552  0.357499  ... -0.833921 -0.136692   

            maxfun   meandom    mindom    maxdom   dfrange   modindx  \
meanfreq  0.274004  0.536666  0.229261  0.519528  0.515570 -0.216979   
sd       -0.129662 -0.482726 -0.357667 -0.482278 -0.475999  0.122660   
median    0.251328  0.455943  0.191169  0.438919  0.435621 -0.213298   
Q25       0.199841  0.467403  0.302255  0.459683  0.454394 -0.141377   
Q75       0.285584  0.359181 -0.023750  0.335114  0.335648 -0.216475   
IQR      -0.069588 -0.333362 -0.357037 -0.337877 -0.331563  0.041252   
skew     -0.080861 -0.336848 -0.061608 -0.305651 -0.304640 -0.169325   
kurt 

The variables that correlate with our new dummy variables, male and female, the most are used for the model. 

In [46]:
df = df[['IQR','sp.ent', 'meanfun', 'sd', 'Q25','centroid','sfm','meanfreq', 'female']] #make this selection before you dropna
df = df.dropna() #first get rid of rows with empty cells 
df['female'].value_counts()

1    1584
0    1584
Name: female, dtype: int64

# Building the Model

For this prediction, we are using the _K nearest neighbor_ algorithm. 

KNN is an algorithm that determines the value of a data point by the data points around it (neighbors). KNN classifier determines the class of a data point by majority voting principle. Basically this means that if k is set to 3, the classes of 3 closest points are checked. Prediction is done according to the majority class. 

In [60]:
from sklearn.preprocessing import normalize #get the function needed to normalize our data.

X = df[['IQR','sp.ent', 'meanfun', 'sd', 'Q25','centroid','sfm','meanfreq']] #create the X matrix
#X = normalize(X) #normalize the matrix to put everything on the same scale

y = df['female'] #create the y-variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) #split the data, store it into different variables
#the order of the above variables is very important
X_train.head()

IQR    sp.ent   meanfun        sd       Q25  centroid       sfm  \
1866  0.174542  0.941751  0.174179  0.088698  0.043235  0.141083  0.601127   
767   0.100853  0.908148  0.100181  0.057011  0.106603  0.167021  0.420416   
2862  0.028535  0.813883  0.179095  0.034413  0.223822  0.234486  0.166600   
1064  0.105827  0.911062  0.138659  0.057609  0.138016  0.197073  0.322959   
270   0.108663  0.940602  0.108932  0.071383  0.094235  0.142018  0.613985   

      meanfreq  
1866  0.141083  
767   0.167021  
2862  0.234486  
1064  0.197073  
270   0.142018

Now we will use the _KNeighborsClassifier_ from sklearn.

In [61]:
from sklearn.neighbors import KNeighborsClassifier #the object class we need

knn = KNeighborsClassifier(n_neighbors=3) #create a KNN-classifier with 3 neighbors
knn = knn.fit(X_train, y_train) #this fits the k-nearest neigbor model with the train data

# Model Evaluation

Now that we have cleaned the dataset, split the dataset into a training and a test set and preformed the KNN algorithm with 3 neighbors. Let's evaluate the results, beginning with accuracy.

In [62]:
knn.score(X_test, y_test) #calculate the fit on the test data

0.9758149316508938

The accuracy is 97% this is very good but does not necessarily tell the whole story of our predictions. Let's take a look at the confusion matrix to get a clearer picture.

In [63]:
from sklearn.metrics import confusion_matrix
y_test_pred = knn.predict(X_test) #the predicted values
cm = confusion_matrix(y_test, y_test_pred) #creates a "confusion matrix"
cm

array([[482,  12],
       [ 11, 446]])

In [64]:
conf_matrix = pd.DataFrame(cm, index=['Not female (actual)', 'Female (actual)'], columns = ['Not female (predicted)', 'Female (predicted)']) 
conf_matrix

Not female (predicted)  Female (predicted)
Not female (actual)                     482                  12
Female (actual)                          11                 446

The way to read this is that of the predicted female voices, 446 are correctly predicted as _female_, 12 voices that were predicted as female were actually male. And of those who were predicted as _not female_, 482 were predicted correctly, while 11 predicted _not female_ voices were actually female voices. The _recall_ and _precision_ for the voice predictions:

$recall = \frac{446}{ 11 + 446} = .98$

$precision = \frac{446 }{ 446 + 12} = .97$

We might improve our scores by trying out different values of _k_.

The recall is 98% which means that 98% of the female voices where correctly predicted to be female voices.

The precision is 97% which means that 97% of the predicted female voices were actually female voices.